File: sturdr/dsp/gnss_signal.py
"""**gnss_signal.py**

======  ============================================================================================
file    sturdr/dsp/gnss_signal.py
brief   Common GNSS signal functions. 
date    October 2024
refs    1. "A Software-Defined GPS and Galileo Receiver: A Single-Frequency Approach", 2007
            - Borre, Akos, Bertelsen, Rinder, Jensen
        2. "Understanding GPS/GNSS Principles and Applications", 3rd Edition, 2017 
            - Kaplan & Hegarty
======  ============================================================================================
"""

import numpy as np
from numba import njit

# ================================================================================================ #

@njit(cache=True, fastmath=True)
def CorrelateEPL(signal: np.ndarray, code: np.ndarray, tap_spacing: int, noise_spacing: bool=False):
    """
    Generate the standard Early-Prompt-Late correlator values

    Parameters
    ----------
    signal : np.ndarray
        Recorded signal data and carrier replica multiplied together
    code : np.ndarray
        Local code replica from NCO
    tap_spacing : int
        Number of samples between each correlator

    Returns
    -------
    IE, QE, IP, QP, IL, QL, IP_1, QP_1, IP_2, QP_2 : np.double
        GNSS correlator values
    """
    len = code.size
    half_len = np.int32(len / 2)
    idx = np.arange(len, dtype=np.int32)
    
    # correlate
    E = np.sum(signal * code[(idx + tap_spacing) % len])
    L = np.sum(signal * code[(idx - tap_spacing) % len])
    # E = np.sum(signal * code[(idx + tap_spacing)])
    # L = np.sum(signal * code[(idx - tap_spacing)])
    p1 = np.sum(signal[:half_len] * code[:half_len])
    p2 = np.sum(signal[half_len:] * code[half_len:])
    P = p1 + p2
    
    if noise_spacing:
        N = np.sum(signal * code[(idx + half_len) % len])
        # N = np.sum(signal * code[(idx + noise_spacing)])
        return E.real, E.imag, P.real, P.imag, L.real, L.imag, p1.real, p1.imag, p2.real, p2.imag, \
               N.real, N.imag
    else:
        # return IE, QE, IP, QP, IL, QL, IP_1, QP_1, IP_2, QP_2
        return E.real, E.imag, P.real, P.imag, L.real, L.imag, p1.real, p1.imag, p2.real, p2.imag, \
               np.nan, np.nan

@njit(cache=True, fastmath=True)
def Correlate(signal: np.ndarray, code: np.ndarray):
    """
    Correlate a signal and code replica

    Parameters
    ----------
    signal : np.ndarray
        Recorded signal data and carrier replica multiplied together
    code : np.ndarray
        Local code replica from NCO

    Returns
    -------
    I, Q : np.double
        GNSS correlator values
    """
    # correlate
    C = np.sum(signal * code)
    
    # return IE, QE, IP, QP, IL, QL, IP_1, QP_1, IP_2, QP_2
    return C.real, C.imag

# ================================================================================================ #

@njit(cache=True, fastmath=True)
def CodeNCO(code: np.ndarray, 
            sampling_freq: np.double, 
            code_freq: np.double, 
            code_len: np.double, 
            remainder_phase: np.double = 0.0):
    """
    Return the upsampled version of the code provided

    Parameters
    ----------
    code : np.ndarray
        Code to upsample
    sampling_freq : np.double
        GNSS receiver front end sampling frequency [Hz]
    code_freq : np.double
        GNSS signal code frequency [Hz]
    code_len : np.double
        Length of the normal (not upsampled) code
    remainder_phase : np.double, optional
        Initial fractional phase of the code, by default 0.0

    Returns
    -------
    upsampled_code : np.ndarray
        Upsampled version of provided code
    remainder_phase : np.double
        Initial fractional phase of the code during the next period
    """
    
    samples_per_code = np.round(sampling_freq / (code_freq / code_len)) + 1
    phase_points = np.mod(remainder_phase + np.arange(samples_per_code) * code_freq / sampling_freq, code_len)
    upsampled_code = code[phase_points[:-1].astype(np.int32)]
    remainder_phase = np.mod(phase_points[-1], code_len)
    
    return upsampled_code, remainder_phase

# ================================================================================================ #

@njit(cache=True, fastmath=True)
def CarrierNCO(sampling_freq: np.double, 
               freq: np.double, 
               freq_jitter: np.double,
               size: np.double, 
               remainder_phase: np.double = 0.0):
    """
    Create a local carrier replica

    Parameters
    ----------
    sampling_freq : np.double
        Sampling frequency of the front end receiver [Hz]
    freq : np.double
        Current carrier frequency [rad/s]
    freq_jitter : np.double
        Current carrier frequency jitter [rad/s^2]
    size : np.double
        Desired replica length
    remainder_phase : np.double, optional
        Initial fractional phase of the carrier, by default 0.0

    Returns
    -------
    replica : np.ndarray
        Local carrier signal replica
    remainder_phase : np.double
        Initial fractional phase of the carrier during the next period
    """
    T = np.arange(size+1) / sampling_freq
    phase_points = (freq + 0.5 * freq_jitter * T) * T + remainder_phase
    replica = np.exp(-1j * phase_points[:-1])
    remainder_phase = np.remainder(phase_points[-1], 2*np.pi)
    
    return replica, remainder_phase

# ================================================================================================ #

@njit(cache=True, fastmath=True)
def shift(arr, num):
    result = np.empty_like(arr)
    if num > 0:
        result[:num] = arr[-num:]
        result[num:] = arr[:-num]
    elif num < 0:
        result[num:] = arr[-num:]
        result[:num] = arr[-num:]
    else:
        result[:] = arr
    return result

# ================================================================================================ #

@njit(cache=True, fastmath=True)
def AccumulateEPL(rfdata: np.ndarray, 
                  code: np.ndarray, 
                  tap_spacing: np.double,
                  sampling_freq: np.double, 
                  code_len: np.double,
                  code_freq: np.double,
                  carrier_freq: np.double,
                  carrier_jitter: np.double,
                  rem_code_phase: np.double, 
                  rem_carrier_phase: np.double,
                  n_samples: int,
                  samples_accumulated: int,
                  half_samples: int,
        ):
    E   = 0.0
    L   = 0.0
    N   = 0.0
    P_1 = 0.0
    P_2 = 0.0
    
    d_carrier = (carrier_freq + 0.5 * carrier_jitter / sampling_freq) / sampling_freq
    d_code    = code_freq / sampling_freq
    noise_tap = 150 # 150 chips from prompt estimate
    
    for i in range(n_samples):
        signal = np.exp(-1j * rem_carrier_phase) * rfdata[i]
        E += (code[np.int32((rem_code_phase + tap_spacing) % code_len)] * signal)
        L += (code[np.int32((rem_code_phase - tap_spacing) % code_len)] * signal)
        N += (code[np.int32((rem_code_phase + noise_tap) % code_len)] * signal)
        if samples_accumulated < half_samples:
            P_1 += (code[np.int32(rem_code_phase % code_len)] * signal)
        else:
            P_2 += (code[np.int32(rem_code_phase % code_len)] * signal)
            
        # propagate
        rem_code_phase += d_code
        rem_carrier_phase += d_carrier
        samples_accumulated += 1
    P = P_1 + P_2
    return E, P, L, P_1, P_2, N, samples_accumulated, rem_carrier_phase, rem_code_phase

File: src/gnss-signal.cpp
/**
 * *gnss-signal.cpp*
 *
 * =======  ========================================================================================
 * @file    sturdr/gnss-signal.cpp
 * @brief   Common GNSS signal functions.
 * @date    January 2025
 * @author  Daniel Sturdivant <sturdivant20@gmail.com>
 * @ref     1. "A Software-Defined GPS and Galileo Receiver: A Single-Frequency Approach", 2007
 *              - Borre, Akos, Bertelsen, Rinder, Jensen
 *          2. "Understanding GPS/GNSS Principles and Applications", 3rd Edition, 2017
 *              - Kaplan & Hegarty
 * =======  ========================================================================================
 */

#include "sturdr/gnss-signal.hpp"

#include <cmath>
#include <navtools/constants.hpp>

namespace sturdr {

// *=== CircShift ===*
Eigen::VectorXcd CircShift(const Eigen::Ref<const Eigen::VectorXcd> &vec, int shift) {
  int n = vec.size();
  shift = shift % n;  // Ensure shift is within bounds

  if (shift == 0) {
    return vec;
  }

  Eigen::VectorXcd shifted(n);

  if (shift > 0) {
    shifted << vec.tail(n - shift), vec.head(shift);
  } else {
    shift = -shift;
    shifted << vec.tail(shift), vec.head(n - shift);
  }

  return shifted;
}

// *=== AccumulateEPL ===*
void AccumulateEPL(
    const Eigen::Ref<const Eigen::VectorXcd> &rfdata,
    const bool code[1023],
    double &rem_code_phase,
    double &code_freq,
    double &rem_carr_phase,
    double &carr_freq,
    double &carr_jit,
    double &samp_freq,
    uint64_t &half_samp,
    uint64_t &samp_remaining,
    double &t_space,
    std::complex<double> &E,
    std::complex<double> &P1,
    std::complex<double> &P2,
    std::complex<double> &L) {
  // init phase increments
  double d_code = code_freq / samp_freq;
  double d_carr = (carr_freq + 0.5 * carr_jit / samp_freq) / samp_freq;

  // loop through number of samples
  std::complex<double> v_carr;
  double v_code;
  for (const std::complex<double> &sample : rfdata) {
    v_carr = std::exp(-navtools::COMPLEX_I<> * rem_carr_phase) * sample;

    // early
    // v_code = code[static_cast<int>(std::fmod(rem_code_phase + t_space, 1023.0))] ? 1.0 : -1.0;
    v_code = code[static_cast<int>(std::round(rem_code_phase + t_space)) % 1023] ? 1.0 : -1.0;
    E += (v_code * v_carr);

    // late
    // v_code = code[static_cast<int>(std::fmod(rem_code_phase - t_space, 1023.0))] ? 1.0 : -1.0;
    v_code = code[static_cast<int>(std::round(rem_code_phase - t_space)) % 1023] ? 1.0 : -1.0;
    L += (v_code * v_carr);

    // prompt
    // v_code = code[static_cast<int>(std::fmod(rem_code_phase, 1023.0))] ? 1.0 : -1.0;
    v_code = code[static_cast<int>(std::round(rem_code_phase)) % 1023] ? 1.0 : -1.0;
    if (samp_remaining > half_samp) {
      P1 += (v_code * v_carr);
    } else {
      P2 += (v_code * v_carr);
    }

    // increment
    rem_code_phase += d_code;
    rem_carr_phase += d_carr;
    samp_remaining--;
  }
}
void AccumulateEPLArray(
    const Eigen::Ref<const Eigen::MatrixXcd> &rfdata,
    const bool code[1023],
    double &rem_code_phase,
    double &code_freq,
    double &rem_carr_phase,
    double &carr_freq,
    double &carr_jit,
    double &samp_freq,
    uint64_t &half_samp,
    uint64_t &samp_remaining,
    double &t_space,
    std::complex<double> E,
    Eigen::Ref<Eigen::VectorXcd> P1,
    Eigen::Ref<Eigen::VectorXcd> P2,
    std::complex<double> L) {
  // init phase increments
  double d_code = code_freq / samp_freq;
  double d_carr = (carr_freq + 0.5 * carr_jit / samp_freq) / samp_freq;

  // loop through number of samples
  int n_samp = rfdata.rows();
  Eigen::VectorXcd v_carr;
  double v_code;
  for (int ii = 0; ii < n_samp; ii++) {
    v_carr = std::exp(-navtools::COMPLEX_I<> * rem_carr_phase) * rfdata.row(ii);

    // early
    // v_code = code[static_cast<int>(std::fmod(rem_code_phase + t_space, 1023.0))] ? 1.0 : -1.0;
    v_code = code[static_cast<int>(std::round(rem_code_phase + t_space)) % 1023] ? 1.0 : -1.0;
    E += (v_code * v_carr(0));

    // late
    // v_code = code[static_cast<int>(std::fmod(rem_code_phase - t_space, 1023.0))] ? 1.0 : -1.0;
    v_code = code[static_cast<int>(std::round(rem_code_phase - t_space)) % 1023] ? 1.0 : -1.0;
    L += (v_code * v_carr(0));

    // prompt
    // v_code = code[static_cast<int>(std::fmod(rem_code_phase, 1023.0))] ? 1.0 : -1.0;
    v_code = code[static_cast<int>(std::round(rem_code_phase)) % 1023] ? 1.0 : -1.0;
    if (samp_remaining > half_samp) {
      P1 += (v_code * v_carr);
    } else {
      P2 += (v_code * v_carr);
    }

    // increment
    rem_code_phase += d_code;
    rem_carr_phase += d_carr;
    samp_remaining--;
  }
}

//* === Correlate ===*
std::complex<double> Correlate(
    const Eigen::Ref<const Eigen::VectorXcd> &rfdata,
    const Eigen::Ref<const Eigen::VectorXcd> &carr,
    const Eigen::Ref<const Eigen::VectorXcd> &code) {
  return (code.array() * carr.array() * rfdata.array()).sum();
}
void Correlate(
    const Eigen::Ref<const Eigen::VectorXcd> &rfdata,
    const bool code[1023],
    double &rem_code_phase,
    double &code_freq,
    double &rem_carr_phase,
    double &carr_freq,
    double &carr_jit,
    double &samp_freq,
    std::complex<double> &C) {
  // init phase increments
  double d_code = code_freq / samp_freq;
  double d_carr = (carr_freq + 0.5 * carr_jit / samp_freq) / samp_freq;

  // loop through number of samples
  std::complex<double> v_carr;
  double v_code;
  for (const std::complex<double> &sample : rfdata) {
    v_carr = std::exp(-navtools::COMPLEX_I<> * rem_carr_phase) * sample;
    v_code = code[static_cast<int>(std::round(rem_code_phase)) % 1023] ? 1.0 : -1.0;
    C += (v_code * v_carr);

    // increment
    rem_code_phase += d_code;
    rem_carr_phase += d_carr;
  }
}

// *=== CodeNCO ===*
Eigen::VectorXcd CodeNCO(
    const bool code[1023],
    const double &code_freq,
    const double &samp_freq,
    double &rem_phase,
    uint64_t &n_samp) {
  // init
  Eigen::VectorXd code_up(n_samp);
  double d_code = code_freq / samp_freq;

  // upsample
  for (uint64_t i = 0; i < n_samp; i++) {
    code_up(i) = code[static_cast<uint64_t>(std::fmod(rem_phase, 1023.0))] ? 1.0 : -1.0;
    rem_phase += d_code;
  }

  return code_up;
}
Eigen::VectorXcd CodeNCO(
    const bool code[1023], const double &code_freq, const double &samp_freq, double &rem_phase) {
  uint64_t n_samp = static_cast<uint64_t>(std::ceil((1023.0 - rem_phase) / code_freq * samp_freq));
  return CodeNCO(code, code_freq, samp_freq, rem_phase, n_samp);
}

// *=== CarrierNCO ===*
Eigen::VectorXcd CarrierNCO(
    const double &carr_freq,
    const double &carr_jit,
    const double &samp_freq,
    double &rem_phase,
    const uint64_t &n_samp) {
  // init
  Eigen::VectorXcd carr_up(n_samp);
  double d_carr = (carr_freq + 0.5 * carr_jit / samp_freq) / samp_freq;

  // upsample
  for (uint64_t i = 0; i < n_samp; i++) {
    carr_up(i) = std::exp(-navtools::COMPLEX_I<> * rem_phase);
    rem_phase += d_carr;
  }

  return carr_up;
}

}  // namespace sturdr